contracts/api3-server-v1/proxies/Api3CrossRateReaderProxyV1.sol

```diff
@@ -0,0 +1,195 @@
+    address public immutable override proxy2;
+
+    /// @notice Calculation type to be used for the cross rate
+    CalculationType public immutable override calculationType;
+
+    /// @notice Derived dAPI name of the cross rate as a bytes32 string
+    /// @dev For example if the two proxies are "USD/USDT" and "USDT/EUR", the
+    /// derived dAPI name should be "USD/EUR". This is just for informational
+    /// purposes and is not used in any calculations
+    bytes32 public immutable override crossRateDapiName;
+
+    /// @dev Parameters are validated by Api3CrossRateReaderProxyV1Factory
+    /// @param proxy1_ First IApi3ReaderProxy contract address
+    /// @param proxy2_ Second IApi3ReaderProxy contract address
+    /// @param calculationType_ Calculation type to be used for the cross rate
+    /// @param crossRateDapiName_ dAPI name of the cross rate as a bytes32 string
+    constructor(
+        address proxy1_,
+        address proxy2_,
+        CalculationType calculationType_,
+        bytes32 crossRateDapiName_
+    ) {
+        proxy1 = proxy1_;
+        proxy2 = proxy2_;
+        calculationType = calculationType_;
+        crossRateDapiName = crossRateDapiName_;
+        _disableInitializers();
+    }
+
+    /// @notice Initializes the contract with the initial owner
+    /// @param initialOwner Initial owner
+    function initialize(address initialOwner) external override initializer {
+        __Ownable_init(initialOwner);
+    }
+
+    /// @notice Returns the current value and timestamp of the cross rate between
+    /// two API3 data feeds associated with the two IApi3ReaderProxy contracts
+    /// @dev The value is calculated based on the calculation type and the
+    /// timestamp is the earliest of the two timestamps to ensure that the data
+    /// is not stale
+    /// @return value Cross rate between the two data feed values
+    /// @return timestamp Timestamp of the oldest data feed update
+    function read()
+        public
+        view
+        override
+        returns (int224 value, uint32 timestamp)
+    {
+        (int224 value1, uint32 timestamp1) = IApi3ReaderProxy(proxy1).read();
+        (int224 value2, uint32 timestamp2) = IApi3ReaderProxy(proxy2).read();
+
+        int256 val1 = int256(value1);
+        int256 val2 = int256(value2);
+        int256 crossRate;
+        if (calculationType == CalculationType.Divide1By2) {
+            if (val2 == 0) revert ProxyReturnedZero(proxy2);
+            crossRate = (val1 * 1e18) / val2;
+        } else if (calculationType == CalculationType.Divide2By1) {
+            if (val1 == 0) revert ProxyReturnedZero(proxy1);
+            crossRate = (val2 * 1e18) / val1;
+        } else if (calculationType == CalculationType.Multiply) {
+            crossRate = (val1 * val2) / 1e18;
+        }
+        if (crossRate > type(int224).max || crossRate < type(int224).min) {
+            revert ValueOutOfRange(crossRate);
+        }
```

Comment:
I'm also now thinking that this check is redundant because there is no way a multiplication that does not overflow `int256` could result in a value that does overflow `int224` after being divided by `1e18`